### PR TITLE
Feature/1178 authenticate apis

### PR DESCRIPTION
**Pull request by jwhite3**
*What does this PR do?*
#60 

Blocked by: 
- [ ] https://github.com/CMSgov/blueprint_ui/pull/62

### CODE REVIEW

File: blueprintapi/urls.py
Comment (rmoore-theta) [inline blueprintapi/urls.py:38]: Was it intentional to use this url instead of conforming to the others by starting with api/?

**Comment by jwhite3** [comment]
To do: 
- [x] Expire user token

File: blueprintapi/tests.py
Comment (jwhite3) [inline blueprintapi/tests.py:10]: This list certainly isn't exhaustive, but it's really only meant to test that without a token, you get an unauthenticated error. Some endpoints don't have these restrictions though.

File: blueprintapi/permissions.py
Comment (jwhite3) [inline blueprintapi/permissions.py:4]: Forgot a Docstring here. I'll add it

File: blueprintapi/urls.py
Comment (jwhite3) [inline blueprintapi/urls.py:38]: I intentionally didn't put the auth token endpoint under the `/api` path, but if folks would prefer that it gets moved there, let me know.

File: blueprintapi/views.py
Comment (jwhite3) [inline blueprintapi/views.py:25]: Need to extend this to expire the token.

File: catalogs/views.py
Comment (jwhite3) [inline catalogs/views.py:18]: `AllowAny` is open permissions, but let me know if this should be read-only.

File: catalogs/views.py
Comment (jwhite3) [inline catalogs/views.py:24]: Same with controls, should this be read-only?

File: components/tests.py
Comment (jwhite3) [inline components/tests.py]: Didn't mean to add this in this PR. I'll remove.

File: components/views.py
Comment (jwhite3) [inline components/views.py:19]: For list views, we need to be sure to add this permissions filter: only allowed objects should be returned for list views.

File: components/views.py
Comment (jwhite3) [inline components/views.py:63]: This isn't ideal, and we should probably revisit this at some point. Is the purpose just to get a set of available component types?

File: projects/tests.py
Comment (jwhite3) [inline projects/tests.py:331]: This was interesting, and I'm actually not quite sure what was going on. As far as I could tell, the contents were the same, but testing for equality was failing due to some of the types in each of these `data` attributes. 

Need to investigate further, but since we were checking content additionally below, I removed this for now.

File: projects/tests.py
Comment (jwhite3) [inline projects/tests.py:394]: This temporary user has no permissions assigned.

File: projects/tests.py
Comment (jwhite3) [inline projects/tests.py:396]: This is how you can either authenticate as a different user for a particular test. If you want to remove authentication, you can set `user` and `token` to `None` to ensure the request is submitted without any credentials.

File: projects/views.py
Comment (jwhite3) [inline projects/views.py]: Need to actually use this in the `post` method. I forgot to add it in.

File: projects/views.py
Comment (jwhite3) [inline projects/views.py:50]: Need to add and utilize `lookup_field` or `lookup_url_kwarg` here.

File: projects/views.py
Comment (jwhite3) [inline projects/views.py]: This needs a serializer so that we can utilize `generics` more, but that's out of scope of this PR.

File: projects/views.py
Comment (jwhite3) [inline projects/views.py:127]: Need to overwrite `list` instead of `get`. I'll update.

File: pyproject.toml
Comment (jwhite3) [inline pyproject.toml:22]: This was being used in `Environment.py`, but was not a project requirement.

File: users/models.py
Comment (jwhite3) [inline users/models.py]: I go back and forth on this one (`...` vs `pass`). We should probably make a distinction for the project about when to use one vs the other. I think I'll put `pass` back for now.

File: components/tests.py
Comment (jwhite3) [reply components/tests.py]: Actually, I realized that that class attribute was never used, so I'm going to go ahead and update.

File: projects/views.py
Comment (jwhite3) [reply projects/views.py:127]: Nevermind. I think this would require a more significant refactor, so I'll hold on this one.

File: blueprintapi/urls.py
Comment (rmoore-theta) [reply blueprintapi/urls.py:38]: I only suggested this for consistency with the other endpoints that the UI calls, but the front end related code already is handling this exception to the rule, so it is fine with me.

File: projects/views.py
Comment (rmoore-theta) [reply projects/views.py]: Would love to talk offline about this so I can understand more :)

File: users/models.py
Comment (rmoore-theta) [reply users/models.py]: This could be a good dev brainstorming discussion to understand the differences :)

File: projects/views.py
Comment (jwhite3) [reply projects/views.py]: Just kidding, this would require updates to the url to utilize. Need to address in a later PR.

File: projects/views.py
Comment (jwhite3) [reply projects/views.py:127]: ~I lied, I fixed it.~

Comment (jwhite3) [reply projects/views.py:127]: Nope, still needs more of a refactor. Excluding this for this PR.

File: projects/views.py
Comment (jwhite3) [reply projects/views.py:50]: This would need an update to the url patterns, so skipping for this PR.

File: blueprintapi/settings.py
Comment (jwhite3) [inline blueprintapi/settings.py:124]: We should figure out what the default should be. Also, this would be great to have as an environment variable, but I want to wait to deal with that until we remove/refactor `Environment.py` and use the `django-environ` plugin.

File: blueprintapi/authentication.py
Comment (jwhite3) [inline blueprintapi/authentication.py:28]: @amyd11 This will delete the user's token if it's "expired" and result in a 401 "Token has expired" response. The front end will have to redirect to the login page, and then once the user logs in again, a new token will be created.

File: blueprintapi/settings.py
Comment (jauschalley) [reply blueprintapi/settings.py:124]: Praise: Thanks for doing this.  I believe the default should be 4 hours but I'll get confirmation on that.  It's definitely good to be merged as is for the time being.  Of course, validating that the expiration is working is more difficult the longer that the token lives.

File: pyproject.toml
Comment (jauschalley) [reply pyproject.toml:22]: Praise: Great catch!

File: catalogs/views.py
Comment (jauschalley) [reply catalogs/views.py:18]: Question:  Is this the temporary work around that we discussed until we follow up with more exacting permissions?

File: blueprintapi/urls.py
Comment (jauschalley) [reply blueprintapi/urls.py:38]: I doubt this is going to hurt anyone's feelings.

**Review by jauschalley — DISMISSED** [review]
Praise:  Thanks for taking on this task being so thorough in your work.  I'm approving with the gentle reminder that this is on hold until the front end work is complete.  Thanks!

File: catalogs/views.py
Comment (jwhite3) [reply catalogs/views.py:18]: Exactly. I wanted to point it out here just in case someone did have a thought before we start with this.

File: blueprintapi/settings.py
Comment (jwhite3) [reply blueprintapi/settings.py:124]: Awesome, let me know. Super easy to change and since we have it as a settings value (which I'm pretty sure makes it configurable for tests) it should be very easy to write a test for it. I just forgot lol. Let me see about adding a test for this. 

I did confirm that it works "manually" since I already had tokens older than that value.

Comment (jwhite3) [reply blueprintapi/settings.py:124]: All good: https://github.com/CMSgov/blueprint_api/pull/66/commits/8d0ff09d0cd2a4e72e36b57372a51274b0ae0a1f

**Review by jauschalley — APPROVED** [review]

File: projects/views.py
Comment (amyd11) [inline projects/views.py:21]: Love that using the generics here cleans up the code and brings consistency in implementation!

**Review by amyd11 — APPROVED** [review]

**Pull request merged**
